use 2-stage EmbeddingSpMDM interface

jspark1105 opened the pull request:
Summary: Use the 2-stage EmbeddingSpMDM interface in D19425982 to reduce the overhead of code cache lookup and lock contention

Test Plan: CI

Differential Revision: D19425987

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 6461f3f9:


* **8/8** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6461f3f96adf6cf5c7722729705a40c4b9ac9e7b).

### :detective: 8 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4266728) pytorch_linux_xenial_py3_clang5_asan_build (1/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86837114) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86837114))
<details>
<summary>
<code>Jan 16 05:27:18 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:29:15: error: cannot initialize object parameter of type 'caffe2::Operator<caffe2::CPUContext>' with an expression of type 'caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true>'</code>
</summary>

```
Jan 16 05:27:18 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/unique_ptr.h:157:17: note: candidate constructor not viable: requires 0 arguments, but 1 was provided 
Jan 16 05:27:18       constexpr unique_ptr() noexcept 
Jan 16 05:27:18                 ^ 
Jan 16 05:27:18 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/unique_ptr.h:181:7: note: candidate constructor not viable: requires 2 arguments, but 1 was provided 
Jan 16 05:27:18       unique_ptr(pointer __p, 
Jan 16 05:27:18       ^ 
Jan 16 05:27:18 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/unique_ptr.h:193:7: note: candidate constructor not viable: requires 2 arguments, but 1 was provided 
Jan 16 05:27:18       unique_ptr(pointer __p, 
Jan 16 05:27:18       ^ 
Jan 16 05:27:18 In file included from /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.cc:1: 
Jan 16 05:27:18 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:29:15: error: cannot initialize object parameter of type 'caffe2::Operator<caffe2::CPUContext>' with an expression of type 'caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true>' 
Jan 16 05:27:18         this, Input(INDICES)); 
Jan 16 05:27:18               ^~~~~ 
Jan 16 05:27:18 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:25:24: note: in instantiation of member function 'caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true>::RunOnDevice' requested here 
Jan 16 05:27:18   USE_SIMPLE_CTOR_DTOR(SparseLengthsFused8BitRowwiseOp) 
Jan 16 05:27:18                        ^ 
Jan 16 05:27:18 /var/lib/jenkins/workspace/c10/util/Registry.h:184:30: note: in instantiation of function template specialization 'caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true>::SparseLengthsFused8BitRowwiseOp<const caffe2::OperatorDef &, caffe2::Workspace *&>' requested here 
Jan 16 05:27:18     return ObjectPtrType(new DerivedType(args...)); 
Jan 16 05:27:18                              ^ 
Jan 16 05:27:18 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.cc:77:1: note: in instantiation of function template specialization 'c10::Registerer<std::__cxx11::basic_string<char>, std::unique_ptr<caffe2::OperatorBase, std::default_delete<caffe2::OperatorBase> >, const caffe2::OperatorDef &, caffe2::Workspace *>::DefaultCreator<caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true> >' requested here 
Jan 16 05:27:18 REGISTER_CPU_OPERATOR( 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4266718) pytorch_linux_xenial_py2_7_9_build (2/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86837194) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86837194))
<details>
<summary>
<code>Jan 16 05:27:43 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:100:26: error: 'kernel64_' was not declared in this scope</code>
</summary>

```
Jan 16 05:27:43            data.template data<std::uint8_t>(), 
Jan 16 05:27:43            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:27:43            indices.template data<std::int32_t>(), 
Jan 16 05:27:43            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:27:43            lengths.template data<int>(), 
Jan 16 05:27:43            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:27:43            weights, 
Jan 16 05:27:43            ~~~~~~~~         
Jan 16 05:27:43            output->template mutable_data<float>()); 
Jan 16 05:27:43            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:27:43 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:100:26: error: 'kernel64_' was not declared in this scope 
Jan 16 05:27:43        success = kernel64_( 
Jan 16 05:27:43                  ~~~~~~~~~^ 
Jan 16 05:27:43            output_size, 
Jan 16 05:27:43            ~~~~~~~~~~~~     
Jan 16 05:27:43            index_size, 
Jan 16 05:27:43            ~~~~~~~~~~~      
Jan 16 05:27:43            data_size, 
Jan 16 05:27:43            ~~~~~~~~~~       
Jan 16 05:27:43            data.template data<std::uint8_t>(), 
Jan 16 05:27:43            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4266723) pytorch_macos_10_13_py3_build (3/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86837439) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86837439))
<details>
<summary>
<code>Jan 15 21:28:23 FAILED: caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o</code>
</summary>

```
Jan 15 21:28:22 ../caffe2/operators/conv_pool_op_base.h:343:29: warning: comparison of integers of different signs: 'int' and 'std::__1::vector<int, std::__1::allocator<int> >::size_type' (aka 'unsigned long') [-Wsign-compare] 
Jan 15 21:28:22       for (int dim = 0; dim < dims.size(); ++dim) { 
Jan 15 21:28:22                         ~~~ ^ ~~~~~~~~~~~ 
Jan 15 21:28:22 ../caffe2/operators/lp_pool_op.cc:131:31: note: in instantiation of member function 'caffe2::ConvPoolOpBase<caffe2::CPUContext>::ComputePads' requested here 
Jan 15 21:28:22   ConvPoolOpBase<CPUContext>::ComputePads({height, width}); 
Jan 15 21:28:22                               ^ 
Jan 15 21:28:22 5 warnings generated. 
Jan 15 21:28:22 [2179/2991] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/lpnorm_op.cc.o 
Jan 15 21:28:23 [2180/2991] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/lstm_unit_op.cc.o 
Jan 15 21:28:23 [2181/2991] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o 
Jan 15 21:28:23 FAILED: caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o  
ing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o -MF caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o.d -o caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o -c ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.cc 
Jan 15 21:28:23 In file included from ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.cc:1: 
Jan 15 21:28:23 ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:163:11: error: no template named 'EmbeddingSpMDMKernelSignature' in namespace 'fbgemm' 
Jan 15 21:28:23   fbgemm::EmbeddingSpMDMKernelSignature<std::uint8_t, std::int32_t>::Type 
Jan 15 21:28:23   ~~~~~~~~^ 
Jan 15 21:28:23 ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:163:70: error: non-friend class member 'Type' cannot have a qualified name 
Jan 15 21:28:23   fbgemm::EmbeddingSpMDMKernelSignature<std::uint8_t, std::int32_t>::Type 
Jan 15 21:28:23                                                                    ~~^ 
Jan 15 21:28:23 ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:163:74: error: expected ';' at end of declaration list 
Jan 15 21:28:23   fbgemm::EmbeddingSpMDMKernelSignature<std::uint8_t, std::int32_t>::Type 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4266730) pytorch_linux_xenial_py3_6_gcc5_4_build (4/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86837472) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86837472))
<details>
<summary>
<code>Jan 16 05:28:30 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:100:26: error: 'kernel64_' was not declared in this scope</code>
</summary>

```
Jan 16 05:28:30                           ^ 
Jan 16 05:28:30 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h: In instantiation of 'bool caffe2::SparseLengthsFused8BitRowwiseOp<Context, with_weights, is_mean>::DoRunWithType() [with IndexType = long int; Context = caffe2::CPUContext; bool with_weights = false; bool is_mean = false]': 
Jan 16 05:28:30 /var/lib/jenkins/workspace/caffe2/core/operator.h:1330:1:   required from 'static bool caffe2::DispatchHelper<caffe2::TensorTypes<FirstType, Types ...>, ExtraArgs ...>::call(Op*, const caffe2::TypeMeta&) [with Op = caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext>; FirstType = long int; Types = {}; ExtraArgs = {}]' 
Jan 16 05:28:30 /var/lib/jenkins/workspace/caffe2/core/operator.h:1330:1:   required from 'static bool caffe2::DispatchHelper<caffe2::TensorTypes<FirstType, Types ...>, ExtraArgs ...>::call(Op*, const caffe2::TypeMeta&) [with Op = caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext>; FirstType = int; Types = {long int}; ExtraArgs = {}]' 
Jan 16 05:28:30 /var/lib/jenkins/workspace/caffe2/core/operator.h:1330:1:   required from 'static bool caffe2::DispatchHelper<caffe2::TensorTypes<FirstType, Types ...>, ExtraArgs ...>::call(Op*, const caffe2::Tensor&) [with Op = caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext>; FirstType = int; Types = {long int}; ExtraArgs = {}]' 
Jan 16 05:28:30 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:28:63:   required from 'bool caffe2::SparseLengthsFused8BitRowwiseOp<Context, with_weights, is_mean>::RunOnDevice() [with Context = caffe2::CPUContext; bool with_weights = false; bool is_mean = false]' 
Jan 16 05:28:30 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.cc:111:1:   required from here 
Jan 16 05:28:30 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:90:26: error: 'kernel32_' was not declared in this scope 
Jan 16 05:28:30        success = kernel32_( 
Jan 16 05:28:30                           ^ 
Jan 16 05:28:30 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:100:26: error: 'kernel64_' was not declared in this scope 
Jan 16 05:28:30        success = kernel64_( 
Jan 16 05:28:30                           ^ 
Jan 16 05:28:30 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o] Error 1 
Jan 16 05:28:30 make[2]: *** Waiting for unfinished jobs.... 
Jan 16 05:28:30 caffe2/CMakeFiles/torch_cpu.dir/build.make:13690: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o' failed 
Jan 16 05:28:30 CMakeFiles/Makefile2:5419: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Jan 16 05:28:30 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Jan 16 05:28:30 Makefile:138: recipe for target 'all' failed 
Jan 16 05:28:30 make: *** [all] Error 2 
Jan 16 05:28:30 -- Building version 1.5.0a0+6461f3f 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4266727) pytorch_xla_linux_xenial_py3_6_clang7_build (5/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86837717) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86837717))
<details>
<summary>
<code>Jan 16 05:29:35 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:29:15: error: cannot initialize object parameter of type 'caffe2::Operator<caffe2::CPUContext>' with an expression of type 'caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true>'</code>
</summary>

```
Jan 16 05:29:35 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/unique_ptr.h:157:17: note: candidate constructor not viable: requires 0 arguments, but 1 was provided 
Jan 16 05:29:35       constexpr unique_ptr() noexcept 
Jan 16 05:29:35                 ^ 
Jan 16 05:29:35 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/unique_ptr.h:181:7: note: candidate constructor not viable: requires 2 arguments, but 1 was provided 
Jan 16 05:29:35       unique_ptr(pointer __p, 
Jan 16 05:29:35       ^ 
Jan 16 05:29:35 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/unique_ptr.h:193:7: note: candidate constructor not viable: requires 2 arguments, but 1 was provided 
Jan 16 05:29:35       unique_ptr(pointer __p, 
Jan 16 05:29:35       ^ 
Jan 16 05:29:35 In file included from /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.cc:1: 
Jan 16 05:29:35 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:29:15: error: cannot initialize object parameter of type 'caffe2::Operator<caffe2::CPUContext>' with an expression of type 'caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true>' 
Jan 16 05:29:35         this, Input(INDICES)); 
Jan 16 05:29:35               ^~~~~ 
Jan 16 05:29:35 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:25:24: note: in instantiation of member function 'caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true>::RunOnDevice' requested here 
Jan 16 05:29:35   USE_SIMPLE_CTOR_DTOR(SparseLengthsFused8BitRowwiseOp) 
Jan 16 05:29:35                        ^ 
Jan 16 05:29:35 /var/lib/jenkins/workspace/c10/util/Registry.h:184:30: note: in instantiation of function template specialization 'caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true>::SparseLengthsFused8BitRowwiseOp<const caffe2::OperatorDef &, caffe2::Workspace *&>' requested here 
Jan 16 05:29:35     return ObjectPtrType(new DerivedType(args...)); 
Jan 16 05:29:35                              ^ 
Jan 16 05:29:35 /var/lib/jenkins/workspace/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.cc:77:1: note: in instantiation of function template specialization 'c10::Registerer<std::__cxx11::basic_string<char>, std::unique_ptr<caffe2::OperatorBase, std::default_delete<caffe2::OperatorBase> >, const caffe2::OperatorDef &, caffe2::Workspace *>::DefaultCreator<caffe2::SparseLengthsFused8BitRowwiseOp<caffe2::CPUContext, false, true> >' requested here 
Jan 16 05:29:35 REGISTER_CPU_OPERATOR( 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4266715) binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build (6/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86837852) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86837852))
<details>
<summary>
<code>Jan 16 05:30:41 /pytorch/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:100:26: error: ‘kernel64_’ was not declared in this scope</code>
</summary>

```
Jan 16 05:30:41            data.template data<std::uint8_t>(), 
Jan 16 05:30:41            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:30:41            indices.template data<std::int32_t>(), 
Jan 16 05:30:41            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:30:41            lengths.template data<int>(), 
Jan 16 05:30:41            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:30:41            weights, 
Jan 16 05:30:41            ~~~~~~~~         
Jan 16 05:30:41            output->template mutable_data<float>()); 
Jan 16 05:30:41            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:30:41 /pytorch/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:100:26: error: ‘kernel64_’ was not declared in this scope 
Jan 16 05:30:41        success = kernel64_( 
Jan 16 05:30:41                  ~~~~~~~~~^ 
Jan 16 05:30:41            output_size, 
Jan 16 05:30:41            ~~~~~~~~~~~~     
Jan 16 05:30:41            index_size, 
Jan 16 05:30:41            ~~~~~~~~~~~      
Jan 16 05:30:41            data_size, 
Jan 16 05:30:41            ~~~~~~~~~~       
Jan 16 05:30:41            data.template data<std::uint8_t>(), 
Jan 16 05:30:41            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4266733) binary_linux_manywheel_2_7mu_cpu_devtoolset7_build (7/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86837983) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86837983))
<details>
<summary>
<code>Jan 16 05:31:01 /pytorch/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:100:26: error: ‘kernel64_’ was not declared in this scope</code>
</summary>

```
Jan 16 05:31:01            data.template data<std::uint8_t>(), 
Jan 16 05:31:01            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:31:01            indices.template data<std::int32_t>(), 
Jan 16 05:31:01            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:31:01            lengths.template data<int>(), 
Jan 16 05:31:01            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:31:01            weights, 
Jan 16 05:31:01            ~~~~~~~~         
Jan 16 05:31:01            output->template mutable_data<float>()); 
Jan 16 05:31:01            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 16 05:31:01 /pytorch/caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:100:26: error: ‘kernel64_’ was not declared in this scope 
Jan 16 05:31:01        success = kernel64_( 
Jan 16 05:31:01                  ~~~~~~~~~^ 
Jan 16 05:31:01            output_size, 
Jan 16 05:31:01            ~~~~~~~~~~~~     
Jan 16 05:31:01            index_size, 
Jan 16 05:31:01            ~~~~~~~~~~~      
Jan 16 05:31:01            data_size, 
Jan 16 05:31:01            ~~~~~~~~~~       
Jan 16 05:31:01            data.template data<std::uint8_t>(), 
Jan 16 05:31:01            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4266716) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (8/8)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86838024) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86838024))
<details>
<summary>
<code>Jan 16 05:31:12 FAILED: caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o</code>
</summary>

```
Jan 16 05:31:08 [2147/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/instance_norm_op.cc.o 
Jan 16 05:31:08 [2148/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/inference_lstm_op.cc.o 
Jan 16 05:31:09 [2149/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/jsd_op.cc.o 
Jan 16 05:31:09 [2150/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/integral_image_op.cc.o 
Jan 16 05:31:09 [2151/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/heatmap_max_keypoint_op.cc.o 
Jan 16 05:31:10 [2152/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/key_split_ops.cc.o 
Jan 16 05:31:11 [2153/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_ops.cc.o 
Jan 16 05:31:12 [2154/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/last_n_window_collector.cc.o 
Jan 16 05:31:12 [2155/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/length_split_op.cc.o 
Jan 16 05:31:12 [2156/3018] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o 
Jan 16 05:31:12 FAILED: caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o  
no-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o -MF caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o.d -o caffe2/CMakeFiles/torch_cpu.dir/operators/lengths_reducer_fused_8bit_rowwise_ops.cc.o -c ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.cc 
Jan 16 05:31:12 In file included from ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.cc:1:0: 
Jan 16 05:31:12 ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:163:11: error: ‘EmbeddingSpMDMKernelSignature’ in namespace ‘fbgemm’ does not name a template type 
Jan 16 05:31:12    fbgemm::EmbeddingSpMDMKernelSignature<std::uint8_t, std::int32_t>::Type 
Jan 16 05:31:12            ^ 
Jan 16 05:31:12 ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:165:11: error: ‘EmbeddingSpMDMKernelSignature’ in namespace ‘fbgemm’ does not name a template type 
Jan 16 05:31:12    fbgemm::EmbeddingSpMDMKernelSignature<std::uint8_t, std::int64_t>::Type 
Jan 16 05:31:12            ^ 
Jan 16 05:31:12 ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h: In member function ‘bool caffe2::SparseLengthsFused8BitRowwiseOp<Context, with_weights, is_mean>::DoRunWithType()’: 
Jan 16 05:31:12 ../caffe2/operators/lengths_reducer_fused_8bit_rowwise_ops.h:73:9: error: ‘kernel32_’ was not declared in this scope 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jspark1105 committed:
use 2-stage EmbeddingSpMDM interface (#32271)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32271

Use the 2-stage EmbeddingSpMDM interface in D19425982 to reduce the overhead of code cache lookup and lock contention.
Fix an issue in sparse_lengths_sum_benchmarks generating empty indices when average length is small like 1.

Test Plan: CI

Reviewed By: dskhudia

Differential Revision: D19425987

fbshipit-source-id: a263fa0ac8aed4eb582b9834b921cb4ea255ea54
